File: js/problem42.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    if (height.length < 3) {
        return 0;
    }
    
    let leftMax = new Array(height.length), rightMax = new Array(height.length);
    
    leftMax[0] = height[0];
    rightMax[rightMax.length - 1] = height[height.length - 1];
    
    for (let i = 1; i < height.length; i++) {
        leftMax[i] = (height[i] > leftMax[i - 1]) ? height[i] : leftMax[i - 1];
    }
    
    for (let i = height.length - 2; i >= 0; i--) {
        rightMax[i] = (height[i] > rightMax[i + 1]) ? height[i] : rightMax[i + 1];
    }

    let res = 0;
    
    for (let i = 1; i < height.length - 1; i++) {
        res += Math.min(leftMax[i], rightMax[i]) - height[i];
    }
    
    return res;
};

File: leetcodeJava/src/main/java/Problem42.java
// 42. Trapping Rain Water

class Solution {
    public int trap(int[] height) {
        if (height.length <= 1) {
            return 0;
        }
        
        int[] maxLeft = new int[height.length];
        int[] maxRight = new int[height.length];
        
        maxLeft[0] = height[0];
        for (int i = 1; i < maxLeft.length; i++) {
            maxLeft[i] = (height[i] > maxLeft[i - 1]) ? height[i] : maxLeft[i - 1];
        }
        
        maxRight[maxRight.length - 1] = height[height.length - 1];
        for (int i = maxRight.length - 2; i >= 0; i--) {
            maxRight[i] = (height[i] > maxRight[i + 1]) ? height[i] : maxRight[i + 1];
        }
        
        int res = 0;
        for (int i = 0; i < height.length; i++) {
            int volumeAtCurrent = Math.min(maxLeft[i], maxRight[i]) - height[i];
            if (volumeAtCurrent > 0) {
                res += volumeAtCurrent;
            }
        }
        
        return res;
    }
}